dpnp/dpnp_iface_manipulation.py

```diff
@@ -245,7 +264,122 @@ def _calc_parameters(a, axis):
     n = a.shape[axis]
     a_shape = list(a.shape)
 
-    return a, a_ndim, order, axis, slobj, n, a_shape
+    if dpnp.is_supported_array_type(obj) and dpnp.is_supported_array_type(
+        values
+    ):
+        usm_type, exec_q = get_usm_allocations([a, obj, values])
+    elif dpnp.is_supported_array_type(values):
+        usm_type, exec_q = get_usm_allocations([a, values])
+    elif dpnp.is_supported_array_type(obj):
+        usm_type, exec_q = get_usm_allocations([a, obj])
+    else:
+        usm_type, exec_q = a.usm_type, a.sycl_queue
+
+    return Parameters(
+        a, a_ndim, order, axis, slobj, n, a_shape, exec_q, usm_type
+    )
+
+
+def _insert_array_indices(parameters, indices, values, obj):
+    """
+    Utility function for ``dpnp.insert`` when indices is an array with
+    multiple elements.
+
+    """
+
+    a, a_ndim, order, axis, slobj, n, a_shape, exec_q, usm_type = parameters
+
+    is_array = isinstance(obj, (dpnp_array, numpy.ndarray, dpt.usm_ndarray))
+    if indices.size == 0 and not is_array:
+        # Can safely cast the empty list to intp
+        indices = indices.astype(dpnp.intp)
+
+    indices[indices < 0] += n
+
+    numnew = len(indices)
+    ind_sort = indices.argsort(kind="stable")
+    indices[ind_sort] += dpnp.arange(
+        numnew, dtype=indices.dtype, sycl_queue=exec_q, usm_type=usm_type
+    )
+
+    a_shape[axis] += numnew
+    old_mask = dpnp.ones(
+        a_shape[axis], dtype=dpnp.bool, sycl_queue=exec_q, usm_type=usm_type
+    )
+    old_mask[indices] = False
+
+    new = dpnp.empty(
+        a_shape,
+        dtype=a.dtype,
+        order=order,
+        sycl_queue=exec_q,
+        usm_type=usm_type,
+    )
+    slobj2 = [slice(None)] * a_ndim
+    slobj[axis] = indices
+    slobj2[axis] = old_mask
+    new[tuple(slobj)] = values
+    new[tuple(slobj2)] = a
+
+    return new
+
+
+def _insert_singleton_index(parameters, indices, values, obj):
+    """
+    Utility function for ``dpnp.insert`` when indices is an array with
+    one element.
+
+    """
+
+    a, a_ndim, order, axis, slobj, n, a_shape, exec_q, usm_type = parameters
+
+    # In dpnp, `.item()` calls `.wait()`, so it is preferred to avoid it
+    # When possible (i.e. for numpy arrays, lists, etc), it is preferred
+    # to use `.item()` on a NumPy array
+    if isinstance(obj, (slice, dpnp_array, dpt.usm_ndarray)):
+        index = indices.item()
+    else:
+        index = numpy.asarray(obj).item()
+
+    if index < -n or index > n:
+        raise IndexError(
+            f"index {index} is out of bounds for axis {axis} " f"with size {n}"
+        )
+    if index < 0:
+        index += n
+
+    # There are some object array corner cases here, that cannot be avoided
```

Comment:
That note was added in numpy in scope of PR 5022. The indention was to support structed arrays and to pass proper dtype here. Do we need do the same, keeping in mind that we are not supporting structed array?
Anyway it seems the comment needs to rephrase to be more clear.